woocommerce/woocommerce-ios
# Update Empty States of Orders

Closes #2215. Closes #1961. Closes #1992. 


## Design

### Processing

Before | After | After Dark
--------|-------|-
![processing-before](https://user-images.githubusercontent.com/198826/81121694-91564e80-8eec-11ea-878b-49f181e08ee4.png) |![processing-after-light](https://user-images.githubusercontent.com/198826/81121692-90bdb800-8eec-11ea-8a78-d61161fa8b60.png) |![processing-after-dark](https://user-images.githubusercontent.com/198826/81121690-8f8c8b00-8eec-11ea-8922-6c67f21f4b0b.png)

### All Orders

Before | After | After Dark
--------|-------|-
![all-before](https://user-images.githubusercontent.com/198826/81121855-eabe7d80-8eec-11ea-89f9-bbaf3d8c04fc.png)|![all-after-light](https://user-images.githubusercontent.com/198826/81121854-ea25e700-8eec-11ea-892d-adfbb0298e30.png)|![all-after-dark](https://user-images.githubusercontent.com/198826/81121847-e7c38d00-8eec-11ea-9f6a-bc71eba6028b.png)

### Search → Pick a Filter (Empty Results)

Before | After | After Dark
--------|-------|-
![filter-before](https://user-images.githubusercontent.com/198826/81121982-2eb18280-8eed-11ea-9d34-4d7ba989b69b.png)|![filter-after-light](https://user-images.githubusercontent.com/198826/81121976-2ce7bf00-8eed-11ea-973a-09466efa9a7f.png)|![filter-after-dark](https://user-images.githubusercontent.com/198826/81121981-2e18ec00-8eed-11ea-84c0-390ac31063f4.png)

### iPad

Processing | All Orders | Search → Filtered
--------|-------|-
![ipad-processing](https://user-images.githubusercontent.com/198826/81122174-949e0a00-8eed-11ea-9ac2-30ce9ee289de.png)|![ipad-all](https://user-images.githubusercontent.com/198826/81122176-95cf3700-8eed-11ea-98cf-d8438a06997d.png)|![ipad-filtered](https://user-images.githubusercontent.com/198826/81122167-92d44680-8eed-11ea-9e9d-56a83c1dba31.png)

## Changes

### EmptyStateVC Configuration

In #2216, we made `EmptyStateVC` _reusable_. 

I made additional changes on `EmptyStateVC` to make it easier to handle the `actionButton`. The previous implementation is a bit difficult to _configure_ when it comes to the `actionButton` handler. This PR changes that so that consumers would need to pass an `enum` in `configure()`:

https://github.com/woocommerce/woocommerce-ios/blob/abb5d53939cf827673a2feb682091dee44218bc7/WooCommerce/Classes/ViewRelated/ReusableViews/EmptyStateViewController/EmptyStateViewController.swift#L222-L230

If consumers pass the `.withLink` option, then the provided `linkURL` would be handled by the `EmptyStateVC` itself. 

### OrdersViewController

The `OrderVC` now uses the `EmptyStateViewController` instead of the `OverlayMessageView`. To support different types of empty states, consumers can pass the `emptyStateConfig` via the `init()`. 

https://github.com/woocommerce/woocommerce-ios/blob/abb5d53939cf827673a2feb682091dee44218bc7/WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift#L95-L97

As an example, here is how the “All Orders” tab was configured:

https://github.com/woocommerce/woocommerce-ios/blob/abb5d53939cf827673a2feb682091dee44218bc7/WooCommerce/Classes/ViewRelated/Orders/OrdersMasterViewController.swift#L98-L110

### EmptyStateVC Styling

Because the `OrdersVC` design needs to show a darker background for the empty state, I made the `EmptyStateVC` style configurable via `init()`:

https://github.com/woocommerce/woocommerce-ios/blob/bb669664aeb1a56bf722b3441f75ae8da4140f7d/WooCommerce/Classes/ViewRelated/ReusableViews/EmptyStateViewController/EmptyStateViewController.swift#L89-L92

The `Style` is just a simple `enum` that hides the implementation:

https://github.com/woocommerce/woocommerce-ios/blob/bb669664aeb1a56bf722b3441f75ae8da4140f7d/WooCommerce/Classes/ViewRelated/ReusableViews/EmptyStateViewController/EmptyStateViewController.swift#L208-L212

## Testing

1. Use a site that has no orders.
2. Navigate to Orders → Processing.
3. Confirm that you can see the “All Orders have been fulfilled.” design. 
4. Navigate to Orders → All Orders.
5. Confirm that you can see the “Waiting for your first order“ design. 
6. Tap on “Learn more”
7. Confirm that the WooCommerce blog is opened. 
8. Navigate to Orders → Search → any filter
9. Confirm that the “We're sorry, we couldn't find...” design is shown. 

## Reviewing 

Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist
